File: src/main/java/dev/arbor/gtnn/api/recipe/PlantCasingCondition.java
package dev.arbor.gtnn.api.recipe;

import com.google.gson.JsonObject;
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.RecipeCondition;
import dev.arbor.gtnn.api.machine.multiblock.ChemicalPlantMachine;
import lombok.Getter;
import lombok.NoArgsConstructor;
import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.util.GsonHelper;
import net.minecraft.util.Mth;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.Map;

@Getter
@NoArgsConstructor
@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class PlantCasingCondition extends RecipeCondition {
    public static final int BRONZE = 1;
    public static final int STEEL = 2;
    public static final int ALUMINIUM = 3;
    public static final int STAINLESS_STEEL = 4;
    public static final int TITANIUM = 5;
    public static final int TUNGSTEN_STEEL = 6;

    public static Map<Integer, String> CASING_TIERS = Map.of(
            BRONZE, "gtnn.recipe.condition.plant_casing.tier.bronze",
            STEEL, "gtnn.recipe.condition.plant_casing.tier.steel",
            ALUMINIUM, "gtnn.recipe.condition.plant_casing.tier.aluminium",
            STAINLESS_STEEL, "gtnn.recipe.condition.plant_casing.tier.stainless_steel",
            TITANIUM, "gtnn.recipe.condition.plant_casing.tier.titanium",
            TUNGSTEN_STEEL, "gtnn.recipe.condition.plant_casing.tier.tungsten_steel"
    );

    public static final PlantCasingCondition INSTANCE = new PlantCasingCondition();

    private int tier = 0;

    public PlantCasingCondition(int tier) {
        this.tier = Mth.clamp(tier, 1, 6);
    }

    @Override
    public String getType() {
        return "chemical_plant_casing_condition";
    }

    @Override
    public Component getTooltips() {
        return Component.translatable(
                "gtnn.recipe.condition.plant_casing.tooltip",
                tier, Component.translatable(CASING_TIERS.get(tier))
        );
    }

    @Override
    public boolean test(@NotNull GTRecipe gtRecipe, @NotNull RecipeLogic recipeLogic) {
        if (recipeLogic.machine instanceof ChemicalPlantMachine chemicalPlantMachine) {
            return chemicalPlantMachine.getCasingTier() >= tier;
        }
        return false;
    }

    @Override
    public RecipeCondition createTemplate() {
        return new PlantCasingCondition();
    }

    @Override
    public JsonObject serialize() {
        JsonObject value = super.serialize();
        value.addProperty("CPTier", tier);
        return value;
    }

    @Override
    public RecipeCondition deserialize(JsonObject config) {
        super.deserialize(config);
        this.tier = GsonHelper.getAsInt(config, "CPTier", 0);
        return this;
    }

    @Override
    public void toNetwork(FriendlyByteBuf buf) {
        super.toNetwork(buf);
        buf.writeInt(tier);
    }

    @Override
    public RecipeCondition fromNetwork(FriendlyByteBuf buf) {
        super.fromNetwork(buf);
        this.tier = buf.readInt();
        return this;
    }
}

File: src/main/java/dev/arbor/gtnn/api/recipe/PlantCasingCondition.kt
package dev.arbor.gtnn.api.recipe

import com.google.gson.JsonObject
import com.gregtechceu.gtceu.api.machine.feature.multiblock.IMultiController
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.RecipeCondition
import dev.arbor.gtnn.block.PlantCasingBlock
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.chat.Component
import net.minecraft.util.GsonHelper


class PlantCasingCondition(var plantCasing: PlantCasingBlock? = null) : RecipeCondition() {
    companion object{
        val INSTANCE: PlantCasingCondition = PlantCasingCondition()
    }

    override fun getType(): String {
        return "chemical_plant_casing"
    }

    override fun getTooltips(): Component {
        return Component.translatable(
            "gtnn.recipe.condition.plant_casing.tooltip",
            plantCasing!!.getTier() + 1,
            plantCasing!!.getName()
        )
    }

    override fun test(recipe: GTRecipe, recipeLogic: RecipeLogic): Boolean {
        val machine = recipeLogic.getMachine()
        return machine is IMultiController && this.plantCasing != null
    }

    override fun createTemplate(): RecipeCondition {
        return PlantCasingCondition()
    }

    override fun serialize(): JsonObject {
        val value = super.serialize()
        value.addProperty("plantCasing", plantCasing!!.getTier())
        return value
    }

    override fun deserialize(config: JsonObject): RecipeCondition {
        super.deserialize(config)
        this.plantCasing = PlantCasingBlock.getByTier(
            GsonHelper.getAsInt(config, "plantCasing", 0)
        )
        return this
    }

    override fun toNetwork(buf: FriendlyByteBuf) {
        super.toNetwork(buf)
        buf.writeInt(plantCasing!!.getTier())
    }

    override fun fromNetwork(buf: FriendlyByteBuf): RecipeCondition {
        super.fromNetwork(buf)
        this.plantCasing = PlantCasingBlock.getByTier(buf.readInt())
        return this
    }
}
